CB-18507 - Prepared Images for upgrade.

vivprabu opened the pull request:
Adding boolean in upgradeV4Response to track what images are prepared for upgrade.

Added image id to message saved to StackStatus when CLUSTER_UPGRADE_PREPARATION_FINISHED status is saved.
Added boolean field to ImageInfoV4Response to indicate if an image id is already prepared.

Added Unit Tests to validate.

lacikaaa reviewed (changes requested):
as we agreed on it, please store the prepared image related info in clustercomponent. don't use statusreason for anything like this

vivprabu commented:
> as we agreed on it, please store the prepared image related info in clustercomponent. don't use statusreason for anything like this

done! can you please review again? @lacikaaa

lacikaaa reviewed (changes requested):
I don't see any cleanup regarding prepared images. After the upgrade is done and the instances are replaced I think the prepared images should be cleaned

vivprabu commented:
> I don't see any cleanup regarding prepared images. After the upgrade is done and the instances are replaced I think the prepared images should be cleaned

@lacikaaa addressed your comments

vivprabu commented on core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/preparation/event/ClusterUpgradePreparationEvent.java:
```diff
@@ -3,27 +3,37 @@
 
     private final Set<ClouderaManagerProduct> clouderaManagerProducts;
 
+    private final String imageId;
```
done

vivprabu commented on core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java:
```diff
@@ -360,4 +367,31 @@ private void validateOperationOnStack(StackView stack, String operationDescripti
         }
     }
+
+    public void updateOrSavePreparedClusterComponent(long stackId, String componentName,
```
done

vivprabu commented on core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java:
```diff
@@ -360,4 +367,31 @@ private void validateOperationOnStack(StackView stack, String operationDescripti
     }
+
+    public void updateOrSavePreparedClusterComponent(long stackId, String componentName,
+            String imageId) throws IOException {
```
done

vivprabu commented on core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterCommonService.java:
```diff
@@ -360,4 +367,31 @@ private void validateOperationOnStack(StackView stack, String operationDescripti
+    }
+
+    private void createAndSaveClusterComponent(long stackId, String componentName, String imageId) {
+        Cluster clusterReference = clusterService.getClusterReference(stackId);
```
done

vivprabu commented on core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java:
```diff
@@ -155,4 +171,18 @@ private boolean determineReplaceVmsParameter(Stack stack, Boolean replaceVms) {
     }
+
+    private List<String> getPreparedImagesList(long stackId) {
+        ClusterComponent clusterComponent = clusterComponentConfigProvider.getComponent(stackId, ComponentType.CLUSTER_UPGRADE_PREPARED_IMAGES,
```
done

vivprabu commented on core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/preparation/ClusterUpgradePreparationActions.java:
```diff
@@ -131,6 +136,12 @@ protected void doExecute(StackContext context, ClusterUpgradePreparationEvent pa
                 ResourceEvent resourceEvent = ResourceEvent.CLUSTER_UPGRADE_PREPARATION_FINISHED;
+                try {
+                    clusterCommonService.updateOrSavePreparedClusterComponent(resourceId,
+                            ResourceEvent.CLUSTER_UPGRADE_PREPARATION_FINISHED.name(), payload.getImageId());
```
done

keyki commented:
@vivprabu could you please rebase your branch to top of the master so the swagger check will also pass ?

lacikaaa commented on core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/ClusterUpgradeActions.java:
```diff
@@ -177,6 +182,8 @@ protected void doExecute(ClusterUpgradeContext context, ClusterUpgradeSuccess pa
                 StatedImage targetImage = getTargetImage(variables);
                 clusterUpgradeService.clusterUpgradeFinished(context.getStackId(), currentImage, targetImage);
                 stackImageService.removeImageByComponentName(context.getStackId(), TARGET_IMAGE);
+                clusterComponentUpdateService.deletePreparedClusterImages(context.getStackId(),
```
you are passing stack id again to a method which expects cluster id

lacikaaa commented on core/src/main/java/com/sequenceiq/distrox/v1/distrox/StackUpgradeOperations.java:
```diff
@@ -162,4 +181,21 @@ private boolean determineReplaceVmsParameter(Stack stack, Boolean replaceVms) {
     }
+
+    private List<String> getPreparedImagesList(long stackId) {
+        List<String> preparedImages = null;
```
this is a bit antipattern to return with null collection. please return with an empty one instead

vivprabu replied:
done

lacikaaa commented on core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java:
```diff
@@ -0,0 +1,78 @@
+    @Inject
+    private ClusterComponentConfigProvider clusterComponentConfigProvider;
+
+    public void updateOrSavePreparedClusterComponent(long stackId, ComponentType componentType,
```
I think `ComponentType componentType` parameter should be removed, as this method works only with this type of component only

vivprabu replied:
done

vivprabu commented:
> @vivprabu could you please rebase your branch to top of the master so the swagger check will also pass ?

rebased

vivprabu commented on core/src/main/java/com/sequenceiq/cloudbreak/service/ClusterComponentUpdateService.java:
```diff
@@ -0,0 +1,77 @@
+            clusterComponentConfigProvider.deleteClusterComponentByClusterIdAndComponentType(clusterId.get(), componentType);
+            LOGGER.debug("Deleted prepared images from Cluster Component for stack id: {}", stackId);
+        }
+    }
```
done
